File: src/main/java/_sortByNumber/_0_100/_41_缺失的第一个正数/Solution.java
package _sortByNumber._0_100._41_缺失的第一个正数;

/**
 * @Description:
 * @Author:boxiao
 * @Data: 2020/6/27 12:08
 */
public class Solution {
    public int firstMissingPositive(int[] nums) {
        int n = nums.length;
        for (int i = 0 ; i < n; i++) {
            while (nums[i] > 0 && nums[i] <= n && nums[i] != nums[nums[i] - 1]) {
                int temp = nums[i];
                nums[i] = nums[nums[i] - 1];
                nums[temp - 1] = temp;
            }
        }
        for (int i = 0; i < n; i++) {
            if (nums[i] != (i+1)) {
                return i+1;
            }
        }
        return n+1;
    }

    public static void main(String[] args) {
        new Solution().firstMissingPositive(new int[]{0,1,2});
    }
}

File: src/main/java/_sortByNumber/_0_100/_41_缺失的第一个正数/Solution.py
'''
41. 缺失的第一个正数

给你一个未排序的整数数组 nums ，请你找出其中没有出现的最小的正整数。

请你实现时间复杂度为 O(n) 并且只使用常数级别额外空间的解决方案。
'''
from typing import List


class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        for i in range(len(nums)):
            while 0 < nums[i] <= len(nums) and nums[i] != nums[nums[i] - 1]:
                tmp = nums[i]
                nums[i] = nums[nums[i] - 1]
                nums[tmp - 1] = tmp
        for i in range(len(nums)):
            if nums[i] != (i + 1):
                return i + 1
        return len(nums) + 1